feat: Support ZIP-213

upbqdn opened the pull request:
## Motivation

- Close #5472.
- Close #5929.
- Close #9605.

## Solution

- Construct coinbase txs as specified in [ZIP 213](https://zips.z.cash/zip-0213) using the [tx builder](https://docs.rs/zcash_primitives/latest/zcash_primitives/transaction/builder/struct.Builder.html) from `zcash_primitives`.
- Based on https://github.com/zcash/librustzcash/pull/1893
- Based on #10191 

### Follow-up Work

- #10170

### PR Checklist

<!-- Check as many boxes as possible. -->

- [x] The PR name is suitable for the release notes.
- [x] The PR follows the [contribution guidelines](https://github.com/ZcashFoundation/zebra/blob/main/CONTRIBUTING.md).
- [ ] The library crate changelogs are up to date.
- [x] The solution is tested.
- [x] The documentation is up to date.

upbqdn committed:
Implement AsRef<[u8; 32]> for Hash

upbqdn committed:
Impl From<&[u8; 32]> for Hash

upbqdn committed:
Update librustzcash dependencies to construct-coinbase-txs branch

upbqdn committed:
Implement TryFrom<&[u8]> for AuthDigest

upbqdn committed:
Add TryFrom<Address> implementation for TransparentAddress

upbqdn committed:
Update deps `zebra-rpc/Cargo.toml`

upbqdn committed:
Make `zebra-chain::transparent::Address` Copy

upbqdn committed:
Reorder imports

upbqdn committed:
Add DefaultRoots::from_coinbase to construct roots from coinbase

upbqdn committed:
Switch to tx builder in zcash primitives

upbqdn committed:
Refactor and expand coinbase transaction generation tests

upbqdn committed:
Remove unused fake_coinbase_transaction function from zip317 module

upbqdn committed:
Remove unused calculate_miner_fee function from get_block_template

upbqdn committed:
Remove unused standard_coinbase_outputs function from get_block_template

upbqdn committed:
Remove unused calculate_default_root_hashes function

The function was no longer referenced and its removal simplifies the
code.

upbqdn committed:
Update block template snapshots

upbqdn committed:
Cosmetics

Replaced variables referencing network upgrades with direct enum usage
for clarity. Reformatted async trait method for consistent indentation.
No functional changes.

upbqdn committed:
Remove unused imports from get_block_template module

upbqdn committed:
Simplify comments

upbqdn committed:
Support shielded coinbase outputs

upbqdn committed:
Remove obsolete TODO comment about internal miner config code

upbqdn committed:
Rename extra_coinbase_data to miner_data

upbqdn committed:
Add support for miner_memo in config and shielded coinbase transactions

upbqdn committed:
Rename variable miner_address to miner_addr in getblocktemplate method

upbqdn committed:
Update snapshots

upbqdn committed:
Remove old transaction builder module and coinbase input constructor

upbqdn committed:
Move error enum doc comment above the attribute and update formatting

upbqdn committed:
Add anyhow, strum, and strum_macros dependencies to Cargo.toml

upbqdn committed:
Add zcash_[transparent, script, primitives] deps

upbqdn committed:
Add thiserror, lazy_static, strum, and strum_macros to dependencies

upbqdn committed:
Add anyhow, strum, and strum_macros to dev-dependencies in Cargo.toml

upbqdn committed:
Remove unused With<Cookie> implementation for HttpRequestMiddleware

The With<Cookie> trait implementation is no longer needed and has been
deleted to simplify the code. There are no functional changes.

upbqdn committed:
Move With trait from zebra-rpc to zebrad and update documentation

The With trait is now defined in zebrad/src/components.rs for broader
reuse. Its documentation has been clarified and slightly restyled. The
original definition was removed from
zebra-rpc/src/server/http_request_compatibility.rs. No semantic changes
to trait behavior were made.

upbqdn committed:
Add predefined miner addresses per network and type for config defaults

upbqdn committed:
Add With<MinerAddressType> implementation for ZebradConfig

upbqdn committed:
Add MinerParams struct and error handling for miner configuration input

upbqdn committed:
Add more error variants and From impls to TransactionError

upbqdn committed:
Update miner_address in v1.9.0-internal-miner.toml test config

upbqdn committed:
Refactor test configs to remove unnecessary Result

upbqdn committed:
Remove obsolete comment about coinbase transaction inputs

upbqdn committed:
Clarify doc comment for coinbase transaction constructor

upbqdn committed:
Refactor builder build call to use Default for empty arguments

upbqdn committed:
Rename `e` to `err`

Simplifies formatting and improves clarity by using {err} syntax.

upbqdn committed:
Remove MinerData  and update coinbase script handling

upbqdn committed:
Refactor coinbase error handling to use TransactionError type

Replaces generic errors with TransactionError::CoinbaseConstruction for
clarity and consistency in coinbase transaction construction failures.

upbqdn committed:
Add regtest coinbase transaction validity test for Zebra blocks

upbqdn committed:
Simplify conversion to Zatoshis

upbqdn committed:
Update to new lrz version

upbqdn commented:
Replaces #9741.

upbqdn committed:
Update deny.toml

upbqdn committed:
Add Coinbase error mapping and remove unused serde_with import

upbqdn committed:
Fix generating `block_commitments_hash`

upbqdn committed:
Add zcash_primitives zip-233 feature to tx_v6 configuration

upbqdn committed:
Fix tx_v6 feature imports and zip233_amount handling in RPC methods

upbqdn committed:
Update tx_v6 feature gating to require nu7 unstable flag

upbqdn committed:
Skip proptest generation for FailedVerification

upbqdn committed:
Remove Zebra mining log function and related references

upbqdn committed:
Refactor coinbase test using ConfiguredFundingStreamRecipient::new_for

upbqdn committed:
Add conversions between Height and i64 types

upbqdn committed:
Add ZcashSerialize implementation for u8 and remove test-specific impl

upbqdn committed:
Update serialization of miner data

upbqdn committed:
Remove unused coinbase script tests and MinerData references

upbqdn committed:
Update block snapshots for coinbase tests

upbqdn committed:
Update TODOs in Cargo.toml and adjust allow-git URLs in deny.toml

upbqdn committed:
Update console-subscriber to version 0.5 in Cargo manifest and lockfile

upbqdn committed:
Integrate lockbox disbursements into transaction funding streams

upbqdn committed:
Move rand dependency to Windows-only target configuration

upbqdn committed:
Make testnet integration test jobs continue on error in CI workflow

zancas commented:
I notice this has conflicts?

arya2 commented on zebra-rpc/src/methods/types/default_roots.rs:
```diff
@@ -42,3 +52,53 @@ pub struct DefaultRoots {
+                .collect(),
+            chain_history_root,
+            auth_data_root,
+            block_commitments_hash: if NetworkUpgrade::current(net, height)
+                == NetworkUpgrade::Heartwood
+                && chain_history_root == block::CHAIN_HISTORY_ACTIVATION_RESERVED.into()
```
I introduced this mistake earlier, I thought your PR had fixed it, but it's still here.

```suggestion
            block_commitments_hash: if NetworkUpgrade::current(net, height) < NetworkUpgrade::Nu5 {
                chain_history_root.bytes_in_serialized_order().into()
```


or 


```rust

// instead of `chain_history_root.bytes_in_serialized_order().into()`,
chain_history_root.bytes_into()

pub trait BytesInDisplayOrder {
..
    /// Converts the bytes in the internal serialized order to the expected type.
    fn bytes_into<T: From<[u8; BYTE_LEN]>>(&self) -> T {
        self.bytes_in_serialized_order().into()
    }
}
```

arya2 replied:
This suggestion still applies, the `block_commitments_hash` field should be the plaing chain history root bytes if `chain_history_root == CHAIN_HISTORY_ACTIVATION_RESERVED && (nu == Heartwood || nu == Canopy)`.

I'm not sure what should happen pre-Canopy.

upbqdn committed:
Remove unused rand crate import from methods module

arya2 commented on zebra-chain/src/transparent/serialize.rs:
🎉

arya2 commented on zebra-rpc/src/methods/types/get_block_template/constants.rs:
```diff
@@ -2,11 +2,15 @@
     parameters::subsidy::FundingStreamReceiver::{self, *},
 };
 
+/// The maximum length of the optional, arbitrary data in the script sig field of a coinbase tx.
```
Optional/Nitpick:

```suggestion
/// The maximum length of the optional, arbitrary data in the script sig field of a coinbase tx that
/// can be configured to be included in the output of calls to the `getblocktemplate` RPC method.
```

arya2 commented on zebra-rpc/src/methods/types/get_block_template.rs:
```diff
@@ -453,86 +545,31 @@ where
 }
 
-// A limit on the configured extra coinbase data, regardless of the current block height.
-// This is different from the consensus rule, which limits the total height + data.
```
The second doc line, 457, would be nice to include for the new constant (instead of my suggestion there).

arya2 commented on zebrad/tests/common/coinbase.rs:
```diff
@@ -0,0 +1,85 @@
+            .wrap_err("possible port conflict with another zebrad instance")
+    }
+
+    try_join_all(MinerAddressType::iter().map(regtest_coinbase))
+        .await
+        .map(|_| ())
```
Nice.

arya2 reviewed (approved):
This looks great ~and good to merge once the dependencies have been published~.

I've pushed a commit with my suggestions, feel free to keep it or make changes.

Great refactors!

Update: This looks like it could merge as-is once the deny lints are fixed.

arya2 commented on Cargo.toml:
```diff
@@ -168,10 +171,27 @@ uint = "0.10"
+zcash_protocol = { git = "https://github.com/zcash/librustzcash", branch = "bump-zcash_script" }
+
+zcash_script = { git = "https://github.com/ZcashFoundation/zcash_script", branch = "make-parse-pub" }
+libzcash_script = { git = "https://github.com/ZcashFoundation/zcash_script", branch = "make-parse-pub" }
```
This seems okay to merge with the dependency patches, these last two are on main, and the others are https://github.com/zcash/librustzcash/compare/main...bump-zcash_script

```suggestion
zcash_script = { git = "https://github.com/ZcashFoundation/zcash_script", branch = "master" }
libzcash_script = { git = "https://github.com/ZcashFoundation/zcash_script", branch = "master" }
```

upbqdn replied:
I don't think we'll be able to release our crates, then.

arya2 replied:
Not until zcash_script and librustzcash have been released and zebra has been updated, but we could do a hot fix release off the prior commit if needed.

upbqdn replied:
That would still block a regular Zebra release, no?

arya2 replied:
Yep. we'd have to revert this commit on main unless we're just publishing a small hotfix. It seems likely enough to me that zcash_script and the librustzcash crates will be published before it's an issue that blocking a regular Zebra release on it would be okay.

upbqdn replied:
We also have patched all the lrz deps. I'll change the branches there to `main`, but still, I'd prefer to wait until we can merge cleanly. I'll mark the PR as ready for review anyway once I push all the fixes.

upbqdn replied:
Actually, I can't point lrz to `main`. I forgot about the transitive dep on `zcash_script`.

upbqdn replied:
Let's keep this open until the new releases are out.

upbqdn commented on zebra-rpc/src/methods/types/default_roots.rs:
```diff
@@ -88,8 +88,8 @@ impl DefaultRoots {
-            block_commitments_hash: if NetworkUpgrade::current(net, height)
-                == NetworkUpgrade::Heartwood
+            // TODO: use final sapling root pre-heartwood (#10334)
+            block_commitments_hash: if NetworkUpgrade::current(net, height) < NetworkUpgrade::Nu5
```
Thanks for the fix. I have a bit more general one in a subsequent PR, so let's keep this one as-is to reduce merge conflicts. We can make additional fixes in that PR, I'll point it out to you.

upbqdn commented on zebra-rpc/src/methods/types/get_block_template/constants.rs:
```diff
@@ -8,7 +8,10 @@ use zebra_chain::{
 
-/// The maximum length of the optional, arbitrary data in the script sig field of a coinbase tx.
+/// The maximum length of the optional, arbitrary data in the script sig field of a coinbase tx that
+/// can be configured to be included in the output of calls to the `getblocktemplate` RPC method.
```
The output of that RPC is an encoded block, and the script sig field mentioned in the comment is part of the coinbase tx's transparent _input_, also encoded as a script, so I'd find it a bit confusing to say that the arbitrary data is in the output of the RPC because then I'd expect to be able to see it there.

upbqdn commented on zebra-rpc/src/methods/types/get_block_template/constants.rs:
```diff
@@ -8,7 +8,10 @@ use zebra_chain::{
+/// The maximum length of the optional, arbitrary data in the script sig field of a coinbase tx that
+/// can be configured to be included in the output of calls to the `getblocktemplate` RPC method.
+///
+/// This is different from the consensus rule, which limits the total height + data.
```
The consensus rule limits the total byte length of the  coinbase tx script (the contents of which might change). That length is the minuend in the formula, so adding that the derived value is different seems a bit confusing.

upbqdn committed:
Remove outdated skip-tree entries for prost-types and hashbrown

upbqdn committed:
Update zcash_script dependencies to use master branch sources

upbqdn commented:
@arya2, the `cargo deny` lints were still failing. I noticed I had them fixed in one of the earlier commits, but reverted those changes in a later commit due to a mistake in rebasing. I fixed the earlier commit and force-pushed. I ended up pushing other fixes that were needed as well, and I replaced your commit. Feel free to submit the suggestions again if you want.

upbqdn committed:
Bump `zcash_script` and don't patch it

upbqdn committed:
Update getrandom skip-tree entry to version 0.3 in deny config

conradoplg commented:
Is this still blocked?

upbqdn committed:
Update Cargo dependencies to use main branch of librustzcash

upbqdn committed:
Upgrade ed25519-zebra to 4.2 and update `deny.toml`

upbqdn committed:
Refactor zip233_amount handling and clean up unused import

upbqdn committed:
Remove transaction builder module from zebra-chain

conradoplg commented on zebra-rpc/src/methods/types/transaction.rs:
```diff
@@ -110,37 +121,151 @@ impl From<VerifiedUnminedTx> for TransactionTemplate<NonNegative> {
+                $res.map_err(|err| tracing::error!("Failed to add {} output: {err}", $type))
+                    .ok()
+            };
+        }
```
Any reason to not just return the error instead of just logging it?

Also, the macro seems simple enough that is might be clearer to just inline the code instead of using it.

upbqdn replied:
I use the macro as a building block in the three closures below, which I then use in this `match` statement, which adds the miner reward to the tx: https://github.com/zcashfoundation/zebra/blob/d2edc5aad48647d234aabf6f3cfdc796223670e0/zebra-rpc/src/methods/types/transaction.rs#L204-L227. The `match` statement tries the ua first, within which the Orchard receiver has the highest precedence, then Sapling, and then the transp one. Along the way, the macros log failures, and the `match` statement returns an error only if it couldn't add the reward to any receiver in the ua, or to the other addr types.

So the `match` statement fails when it couldn't add the reward to the tx at all, and logs the failures implicitly along the way.

I wrote it this way to keep the logic in the `match` statement concise.

upbqdn replied:
But I could get rid of the macro if you prefer that. Here's a version without it: https://github.com/zcashfoundation/zebra/blob/70fcbac23cc3b3231e2ca5340c6dc38a702286fb/zebra-rpc/src/methods/types/transaction.rs#L166-L220.

conradoplg commented on Cargo.toml:
```diff
@@ -23,15 +23,15 @@ license = "MIT OR Apache-2.0"
 homepage = "https://zfnd.org/zebra/"
 keywords = ["zebra", "zcash"]
-rust-version = "1.85.0"
+rust-version = "1.88"
```
Do you remember why this was bumped? If it was because of the LRZ crates then it should be fine

upbqdn replied:
It was because of this: https://github.com/ZcashFoundation/zebra/actions/runs/24003946911/job/70004474592?pr=10048#step:6:408.

upbqdn replied:
The bump is not needed anymore.

conradoplg commented on zebra-rpc/src/config/mining.rs:
```diff
@@ -16,15 +20,18 @@ pub struct Config {
+    ///
+    /// If this string is hex-encoded, it will be hex-decoded into bytes. Otherwise, it will be
+    /// UTF-8 encoded into bytes.
+    pub miner_data: Option<String>,
```
Renaming this requires a major bump in Zebra itself. Should we stick with `extra_coinbase_data`?

upbqdn replied:
Reverted in [ed85a99](https://github.com/ZcashFoundation/zebra/pull/10048/commits/ed85a99196eab2ec3c4cd614cc37f8191fe839c3). Thanks for noticing, the reversion was on my to-do list for this PR.

conradoplg reviewed (commented):
Not a thorough review, added some comments

upbqdn committed:
Update OpenTelemetry and tracing dependencies to latest versions

upbqdn committed:
Move trace_err macro definition below feature-gated zip233 logic

upbqdn commented:
One detail I don't like about this PR is that the core logic for creating coinbase transactions takes place in the `zebra-rpc` crate.

upbqdn committed:
Adjust Input arbitrary data length range to start at 1 instead of 0

upbqdn committed:
Add function to randomize miner data and update coinbase logic

upbqdn committed:
Rename `miner_data` to `extra_coinbase_data` in mining config

upbqdn committed:
Add helper function for fetching default miner addresses

upbqdn committed:
Update changelogs

Zk-nd3r commented:
Running Zebra 4.3.0 + s-nomp on mainnet with shielded coinbase.  Happy to test this branch against our pool when it's ready for external testing,  we consume GBT directly and would catch regressions in coinbase output that break pool software.

One thing we've seen,  s-nomp assumes specific field ordering in the GBT response.  If coinbase tx structure changes shape,  blocks get rejected silently by the network and the pool doesn't know until you dig through Zebra logs.

upbqdn commented:
It is ready for external testing; I've mined a block on Mainnet using this branch. Feel free to let us know about any regressions.

Zk-nd3r commented:
Will build from this branch and point our s-nomp instance at it.  Testing shielded coinbase with the new coinbase construction path.  Will report back with results.

upbqdn commented:
@Zk-nd3r, you're actually better off going with PR #10170. It's in draft because it needs to be merged after this PR, but it's also ready. It targets `main`, but it's based on top of this PR.

conradoplg commented on zebra-chain/src/block/height.rs:
```diff
@@ -148,6 +148,14 @@ impl TryFrom<u32> for Height {
+    type Error = core::num::TryFromIntError;
+
+    fn try_from(height: i64) -> Result<Self, Self::Error> {
+        u32::try_from(height).map(Height)
```
(found with Claude) this does not enforce the Height::MAX restriction. Might be `u32::try_from(height)?.try_into()` instead

upbqdn replied:
That enforcement is actually not needed. I was thinking I'd remove the upper limit entirely. The spec says: 

> Implementations MUST support block heights up to and including 2³¹−1.

That doesn't say anything about the upper limit. I think we should remove the `Height::MAX` limit entirely.

I'd prefer to keep the conversion you referenced as-is.

upbqdn replied:
The type that represents heights in LRZ has no upper limit either.

conradoplg replied:
Makes sense, thanks

conradoplg commented on zebra-chain/src/transparent/serialize.rs:
I'm double-checking this Claude finding but posting here to speed things up. Might be good to repurpose the old tests if possible.

---

zebra-chain/src/transparent/serialize.rs now delegates BIP-34 height parsing to
  zcash_script::opcode::PossiblyBad::parse + PushValue::restrict. The previous in-tree implementation had explicit
  checks that:
  - rejected non-minimal encodings (e.g. 0x01 0x01 for height 1 — must be 0x51),
  - rejected heights outside the valid range for each prefix length,
  - rejected heights > Height::MAX.

  The tests that pinned these (parse_coinbase_height_mins in zebra-chain/src/transparent/tests/vectors.rs, plus the
  round-trip proptests in tests/prop.rs) were deleted with the old code, with no replacement tests against
  Input::zcash_deserialize. If PushValue::restrict does not enforce minimal-encoding identically to zcashd, a
  malicious or buggy block parses successfully on Zebra and mismatches zcashd → chain split. Every node validating
  blocks (mining or not) is affected. Add direct regression tests against Input::zcash_deserialize covering each
  height range and each non-minimal encoding before this can land.

upbqdn replied:
This is a great finding. Fixed in [206d491](https://github.com/ZcashFoundation/zebra/pull/10048/commits/206d491878fdac52d5d360f9e8bab1e961631cd4).

conradoplg reviewed (changes requested):
Some Claude findings that seem relevant

upbqdn committed:
fix(zebra-chain): enforce BIP-34 canonical encoding when parsing coinbase heights

The previous implementation in `Input::zcash_deserialize` parsed coinbase
script-sigs by calling `zcash_script::opcode::PossiblyBad::parse` followed by
`PushValue::restrict`. `restrict` only narrows opcode types (Good vs Bad,
PushValue vs Control vs Operation) — it does not enforce BIP-34 minimal
encoding. As a result, peer-supplied coinbases with non-minimal height pushes
(e.g. height 1 encoded as `0x01 0x01` or `0x04 0x01 0x00 0x00 0x00` instead
of `OP_1`) were accepted by Zebra but would be rejected by `zcashd`, creating
a chain-split risk. The pinning regression tests (`parse_coinbase_height_mins`,
`coinbase_script_round_trip`) had also been removed without replacement.

Replace the parser with a single-oracle encode-and-compare check: read a
candidate height directly off the wire (using the prefix shape to locate the
bytes), then re-encode it via `zcash_script::pattern::push_num` — the same
primitive `zcash_transparent::bundle::TxIn::coinbase` uses to construct
coinbase inputs — and require byte-exact equality. This delegates all
canonicality rules to the upstream encoder. Any non-canonical input (wrong
shape, non-minimal length, oversize push, negative, signed-bit games) fails
the comparison.

Add regression tests at the `Input::zcash_deserialize` boundary:

- `coinbase_deserialize_rejects_non_minimal_heights`: 7 non-minimal cases
  for heights 1, 16, and 17 that the previous code accepted.
- `coinbase_deserialize_accepts_canonical_heights`: boundary heights
  including 1, 16, 17, 127, 128, 32767, 32768, 8388607, 8388608, Height::MAX.
- `coinbase_deserialize_rejects_out_of_range_heights`: sign-bit set, 6-byte
  push, height greater than `Height::MAX`.
- `coinbase_canonical_round_trip`: proptest over the full valid height range.

upbqdn committed:
Upgrade toml from 0.9 to 1.0 to eliminate duplicate crates

Unifies the workspace toml dependency with the version used by the
config crate, removing duplicate toml, toml_datetime, and winnow
entries from the dependency tree.

# Conflicts:
#	Cargo.lock

upbqdn committed:
fix(zebra-rpc): rebuild zip-213 on top of librustzcash 2026-04 cohort

The 2026-04 cohort migration (PR #10522) made the
`zebra_consensus::groth16::SAPLING` re-export private when Sapling moved
to `sapling-crypto::BatchValidator`. Use `LocalTxProver::bundled()`
directly in the V5 transaction template builder, matching how
upstream consumers construct a Sapling prover.

Also drop the stale `[patch.crates-io]` table that pinned librustzcash
crates to the `main` branch — the 2026-04 cohort consumes published
crates.io versions, so the patch is no longer needed.

upbqdn committed:
fix(deps): pin transitive deps to keep MSRV at 1.85.1

Recent dependency releases bumped their MSRV beyond 1.85.1:

  - darling@0.23.0, serde_with@3.18.0 (require 1.88)
  - home@0.5.12 (requires 1.88)
  - icu_*@2.2.0, idna_adapter@1.2.2 (require 1.86)

Pin them in Cargo.lock to the last MSRV-compatible versions via
'cargo update --precise', without touching Cargo.toml or bumping the
workspace MSRV:

  cargo update --precise 3.17.0 -p serde_with
  cargo update --precise 0.5.11 -p home
  cargo update --precise 2.5.4 -p url
  cargo update --precise 1.0.3 -p idna
  cargo update --precise 1.2.0 -p idna_adapter

conradoplg reviewed (commented):
Looks good, thanks! 🎉 

Some tests are failing but seems unrelated, I'm re-running them to see what happens.

Also needs solving conflicts.

Other than that this is now blocked on just deciding if we want to include it in the next release or not.

oxarbitrage committed:
Merge remote-tracking branch 'origin/main' into zip-213

# Conflicts:
#	Cargo.lock
#	deny.toml

conradoplg committed:
remove continue-on-error from workflow to make it work

oxarbitrage reviewed (dismissed):
Reviewed the last commit (dd302dc). Correct — removes continue-on-error from sync-full-testnet and generate-checkpoints-testnet jobs in the integration tests workflow. This was already fixed in main via #10264 but got reverted during the merge into zip-213. Trivial fix, matches the already-merged PR.

conradoplg committed:
Merge branch 'main' of https://github.com/ZcashFoundation/zebra into zip-213

conradoplg committed:
adjust merged test for the new API

oxarbitrage reviewed (approved):
Last two commits look good: merge from main (81ce9f5) and the test fix (e1aba60) adjusting the chain_value_pool_change test to use Input::Coinbase struct directly after Input::new_coinbase() was removed by this PR. Should fix the clippy/unused-deps/check-cargo-lock CI failures.
